File: astexpr/astexpr_java/src/com/iso/astexpr/ASTInterpreter.java
package com.iso.astexpr;

public class ASTInterpreter
{
    private SymbolTable symbolTable;

    public ASTInterpreter()
    {
        symbolTable = new SymbolTable();
    }

    private double evaluate(ExpressionNode node)
    {
        if (node.getValue() != 0)
            return node.getValue();

        if (node.getType() == ExpressionNodeType.PLUS)
        {
            return evaluate((ExpressionNode)node.getLeft()) + evaluate((ExpressionNode)node.getRight());
        }

        if (node.getType() == ExpressionNodeType.MINUS)
        {
            return evaluate((ExpressionNode)node.getLeft()) - evaluate((ExpressionNode)node.getRight());
        }

        if (node.getType() == ExpressionNodeType.DIV)
        {
            return evaluate((ExpressionNode)node.getLeft()) / evaluate((ExpressionNode)node.getRight());
        }

        if (node.getType() == ExpressionNodeType.MUL)
        {
            return evaluate((ExpressionNode)node.getLeft()) * evaluate((ExpressionNode)node.getRight());
        }

        return 0.0;
    }

    public void printExpressionNodes(ExpressionNode node)
    {
        if (node == null) return;

        if (node.getType() != ExpressionNodeType.UNDEFINED)
        {
            System.out.println(node.getType().toString());
            System.out.println("/  \\");
        }
        else
        {
            System.out.println(Integer.toString(node.getValue()) + " ");
        }

        printExpressionNodes((ExpressionNode)node.getLeft());
        printExpressionNodes((ExpressionNode)node.getRight());
    }

    public double interpret(ASTNode node)
    {
        if (node == null) return 0.0;

        if (node.getNodeType() == ASTNodeType.Assignment)
        {
            String variableName = ((VariableNode)node.getLeft()).getName();
            if (!symbolTable.containsKey(variableName)) symbolTable.put(variableName, 0);

            if (node.getRight().getNodeType() == ASTNodeType.Expression)
                symbolTable.put(variableName,evaluate((ExpressionNode)node.getRight()));

            return (double)symbolTable.get(variableName);
        }

        if (node.getNodeType() == ASTNodeType.Expression)
        {
            return evaluate((ExpressionNode)node);
        }

        return 0.0;
    }

}

File: astexpr/astexpr_cpp/astexpr/astinterpreter.cpp
#include "astinterpreter.h"


ASTInterpreter::ASTInterpreter(PASTNode node):
	root(node)
{

}

ASTInterpreter::~ASTInterpreter()
{
	hashTable.clear();
	if (root != nullptr) delete root;
}

float ASTInterpreter::Evaluate(PExpressionNode node)
{
	if (node->GetValue() != 0)
		return node->GetValue();

	if (IsNodeTypeEqual(node, ExpressionNodeType::PLUS))
	{
		return Evaluate(dynamic_cast<PExpressionNode>(node->GetLeft())) +
			Evaluate(dynamic_cast<PExpressionNode>(node->GetRight()));
	}

	if (IsNodeTypeEqual(node, ExpressionNodeType::MINUS))
	{
		return Evaluate(dynamic_cast<PExpressionNode>(node->GetLeft())) -
			Evaluate(dynamic_cast<PExpressionNode>(node->GetRight()));
	}

	if (IsNodeTypeEqual(node, ExpressionNodeType::DIV))
	{
		// check division by zero ?
		return Evaluate(dynamic_cast<PExpressionNode>(node->GetLeft())) /
			Evaluate(dynamic_cast<PExpressionNode>(node->GetRight()));
	}

	if (IsNodeTypeEqual(node, ExpressionNodeType::MUL))
	{
		return Evaluate(dynamic_cast<PExpressionNode>(node->GetLeft())) *
			Evaluate(dynamic_cast<PExpressionNode>(node->GetRight()));
	}

	return 0.0f;
}

bool ASTInterpreter::IsNodeTypeEqual(PExpressionNode node, ExpressionNodeType nodeType)
{
	return node->GetType() == nodeType;
}

float ASTInterpreter::Interpret()
{
	if (root == nullptr) return 0.0f;
 
	if (root->GetNodeType() == ASTNodeType::Assignment)
	{
		string variableName = dynamic_cast<PVariableNode>(root->GetLeft())->GetName();
		if (hashTable.find(variableName) == hashTable.end())
			hashTable[variableName] = 0.0f;

		if (root->GetRight()->GetNodeType() == ASTNodeType::Expression)
			hashTable[variableName] = Evaluate(dynamic_cast<PExpressionNode>(root->GetRight()));

		return hashTable[variableName];
	}

	if (root->GetNodeType() == ASTNodeType::Expression)
	{
		return Evaluate(dynamic_cast<PExpressionNode>(root));
	}

	return 0.0f;
}
